pkg/extendapi/devicetemplate.go

```
@@ -0,0 +1,235 @@
+package extendapi
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/cnrancher/edge-api-server/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+
+	// appsv1 "k8s.io/api/apps/v1"
+	yaml2 "github.com/ghodss/yaml"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	octopusSystemNamespace string = "octopus-system"
+	version                string = "v1alpha1"
+	group                  string = "edge.cattle.io"
+	deviceGroup            string = "devices.edge.cattle.io"
+)
+
+type CreateTemplateRequest struct {
+	DeviceType         string `json:"deviceType"`
+	User               string `json:"user"`
+	AdaptorType        string `json:"adaptorType"`
+	TemplateName       string `json:"templateName"`
+	TemplateDataBase64 string `json:"templateDataBase64"`
+}
+
+type CreateDeviceTemplateHealthHandler struct {
+	clientset *kubernetes.Clientset
+	dyclient  dynamic.Interface
+}
+
+func NewCreateDeviceTemplateHandler(client *kubernetes.Clientset, dyclient dynamic.Interface) *CreateDeviceTemplateHealthHandler {
+	return &CreateDeviceTemplateHealthHandler{
+		clientset: client,
+		dyclient:  dyclient,
+	}
+}
+
+func (h *CreateDeviceTemplateHealthHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		logrus.Error("CreateDeviceTemplateHealthHandler request body error:", err.Error)
+		h.WriteResponse(w, err.Error())
+		return
+	}
+
+	var request CreateTemplateRequest
+
+	if err := json.Unmarshal(body, &request); err != nil {
+		logrus.Error("CreateDeviceTemplateHealthHandler request body validate json error")
+		h.WriteResponse(w, "request body error")
```

the error message is not meaningful, should return the error.Error instead.